docs/DeveloperGuide.md

```diff
@@ -1,38 +1,726 @@
+   Arbitrary way (using index) is not preferred as it is prone to developer erros.
+4. **Easy way to validate user input**  
+   There should also be built-in ways to easily validate components of user input for a command,
+   such as checking length.
+
+#### CommandParser Syntax
+
+The command parser defines any arbitrary user input to be valid
+if it follows the following structure.
+
+```
+mainCommand [payload] [--argument1 [payload1] --argument2 [payload2] ... ]
+```
+
+This should be familiar to you. It is similar to how most CLI applications process arguments in the wild.
+
+![Example](diagrams/git_command.png)
+<figcaption align="center">Example of CLI input syntax in the wild</figcaption>
+
+This achieves design consideration (1). Why?  
+This syntax is intuitive at a glance to our target users,
+is predictable and easy to remember as the only thing they need to remember is the argument name and
+the '--' delimiter.
+
+From this syntax, we can generalise ALL user inputs as `(argument, payload)` pairs.  
+`mainCommand` is a special `argument`, where it MUST be the first word in the user input.
+
+Due to the unique one-to-one relationship between arguments and payloads, we can model a user input
+using this syntax using a `HashMap` mapping each `argument` to a `payload`.
+
+For example,
+`
+$ foo bar --arg1 payload1 payload1--1 --arg2 payload2 --arg3
+`
+
+Will be mapped as:
+`
+(foo, bar), (arg1 payload1 payload1--1), (arg2 payload2), (arg3, "")
+`
+where `""` represents an empty string (for visualization).
+
+Using a `HashMap` fulfils design considerations (2), (3) and (4).
+
+- (2): Order of arguments do not matter as
+- (3): To get a `payload`, the developer simply needs to call `myHashMap.get("argument")`.  
+  This syntatic sugar prevents developer errors compared to an index-based approach.
+- (4): Validating commands is much less difficult using `HashMap`. For example, size can be checked with
+  built-in `.size()`,
+  argument existence can be queried with `.containsKey()`.
+
+#### Integration with WellNUS++
+
+![Integration](diagrams/CommandParserClass.png)
+
+`CommandParser` integrates into the boilerplate via the abstract Manager class.  
+All features are controlled by a manager subclass - hence the developers just need to call
+`getCommandParser` to get a reference to the `CommandParser` taking care of all commands
+in the `Manager` subclass.
+
+#### CommandParser API
+
+There are only two methods that developers need to know to use `CommandParser`.
+
+1. `parseUserInput`
+2. `getMainArgument`
+
+**Usage: `parseUserInput`**
+
+`parseUserInput` is used to get a `HashMap` representation of the user input, a bijection
+between `argument` and `payload` pairs.
 
-{Describe the design and implementation of the product. Use UML diagrams and short code snippets where applicable.}
+Implementation of `parseUserInput`:
 
+![CommandParser implementation](diagrams/CommandParserSequence.png)
```

Comment:
![image](https://user-images.githubusercontent.com/88386677/228426028-93acd900-f0f3-4e02-b46a-f1484675f0a8.png)
A reference frame for the CommandParser accessing its own methods inside the loop may help to better describe what is happening in the sequence diagram?